### DISCUSSION TITLE
Any plans for noRestrictedImports feature parity?

### DISCUSSION THREAD

**Discussion by terrymun** [biomejs/biome#3216] [Q&A] [answered]
While migrating from eslint to biome, I discovered that the `noRestrictedImports` rule only partially implements the feature of `no-restricted-imports`.

- It is not possible to only disallow specific [import names](https://eslint.org/docs/latest/rules/no-restricted-imports#importnames)
- It is not possible to specify [patterns](https://eslint.org/docs/latest/rules/no-restricted-imports#patterns) instead of paths

Are there any plans to update this rule so that the aforementioned features are supported?

**Comment by ematipico** [comment]
Do you mean feature _parity_? 

A user submitted a PR to implement some of the features: https://github.com/biomejs/biome/pull/2977

Other than that, I would ask @anonrig more details because he's the person that implemented the rule.

We're also debating on the naming, and if the rule makes sense, because we also have `useImportRestrictions`

**Comment by feedm3** [comment]
I just tried biome in one of our larger projects and I'm also missing this! Our 2 use cases - maybe someone can point me to an alternative rule to solve the problem:

1. We don't want to allow deep imports of a package, only root-level imports. Our current rule looks like this:
```
'no-restricted-imports': [
      'error',
      {
        patterns: [
          {
            group: ['@example/lib-shared-ui/*'],
            message:
              'Only import from the package root or exposed subfolders. Importing from src or dist is not allowed.',
          },
        ],
      },
    ],
```
2. We have a component lib, and functions form `styled-components` should only be used from that component lib, not `styled-components` itself:
```
'no-restricted-imports': [
      'error',
      {
        paths: [
          {
            name: 'styled-components',
            importNames: ['css', 'styled', 'createGlobalStyle', 'isStyledComponent', 'useTheme', 'keyframes'],
            message: 'Only import from Component Lib as we get into version conflicts otherwise.',
          },
        ]
      },
    ],
```

**Comment by kkimdev** [comment]
+1 pattern & importNames are essential features of this rule.

**Reply by terrymun** [reply]
AFAIK Biome v2's `lint/style/noRestrictedImports` only accept plain strings for paths, and does not seem to support patterns at the moment.

**Comment by jozan** [comment]
in addition to `importNames` i would like to see this rule support dynamic imports as well. eslint doesn't support dynamic imports which is bit of a nuisance.

code
```typescript
import { Clipboard } from 'react-native'; // eslint errors
const { Clipboard } = await import('react-native'); // eslint doesn't error
```

eslint rule
```json
{
  "@typescript-eslint/no-restricted-imports": [
    "error",
    {
      "paths": [
        {
          "name": "react-native",
          "importNames": ["Clipboard"],
          "message": "Don't use deprecated modules from 'react-native'.",
        },
      ],
    },
  ]
}
```

biome's `noRestrictedImports` rule could look like this

```json
{
    "noRestrictedImports": {
        "options": {
            "paths": {
                "react-native": {
                    "importNames": ["Clipboard"],
                    "message": "Don't use deprecated modules from 'react-native'."
                }
            }
        }
    }
}
```

**Comment by quezak** [comment]
+1, we also use the eslint `no-restricted-imports` rule to forbid importing specific symbols from specific packages, not whole packages

**Reply by Conaclos** [reply]
I have started some design work, you can expect parity for the next Biome release. Please consider sponsoring us to make this happen.

**Reply by arendjr** [reply]
@Conaclos is this tracked somewhere? If it doesn't conflict with your work, I might be tempted to at least resurrect https://github.com/biomejs/biome/pull/2977 and get those features in, maybe together with the `importNames` option.

**Reply by Conaclos** [reply]
I didn't open any issue. I planned to rethink the rule completely. We could use the new `RestrictedGlob` for path matching.

**Reply by marcocondrache** [reply]
Could we open an issue to track the progress? I believe this discussion does not provide sufficient tracking for the feature, such as including it in umbrella issues.

**Reply by arendjr** [reply]
I've created a task for adding `patterns`: https://github.com/biomejs/biome/issues/5289

Note that `importNames`/`allowImportNames` are already implemented and will be available in v2.

**Comment by arendjr** [comment, marked as answer]
I've created a task for adding patterns: https://github.com/biomejs/biome/issues/5289

`importNames`/`allowImportNames` have been implemented and will be available in v2.

**Reply by terrymun** [reply]
This is absolutely amazing news! Thank you 😊
